src/explore-education-statistics-common/src/components/Gate.tsx

```diff
@@ -1,61 +1,91 @@
-import useAsyncCallback, {
-  AsyncStateSetterParam,
-} from '@common/hooks/useAsyncCallback';
+import usePrevious from '@common/hooks/usePrevious';
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
-import React, { ReactNode, useEffect } from 'react';
+import React, { ReactNode, useEffect, useState } from 'react';
 
 interface GateProps {
-  condition: boolean | (() => Promise<boolean>);
+  condition: boolean | (() => Promise<boolean>) | (() => boolean);
   children: ReactNode;
+  /**
+   * Fallback to render if the {@property condition}
+   * has not passed yet.
+   */
   fallback?: ReactNode | ((error?: unknown) => ReactNode);
+  /**
+   * Loading state to render if the {@property condition}
+   * is dependent on an asynchronous task.
+   */
   loading?: ReactNode;
+  /**
+   * If the {@property condition} passes once, then we
+   * consider the gate 'closed' and will not try to
+   * re-run the condition and will not unmount the
+   * children if the condition changes back to false.
+   */
+  passOnce?: boolean;
 }
 
 /**
- * Conditionally render {@param children} based on a
- * {@param condition} that must be passed. This may be
- * an asynchronous task or just any boolean.
- *
- * @param loading state can be rendered if it is asynchronous.
- * @param fallback will be rendered if the condition
- * fails or there is an error.
+ * Conditionally render its children based on a condition
+ * that must be passed. This may be an asynchronous task
+ * or just any boolean.
  */
 const Gate = ({
   children,
   condition,
   fallback = null,
   loading = null,
+  passOnce = true,
 }: GateProps) => {
-  const [{ isLoading, value, error }, checkCondition] = useAsyncCallback(
-    async () => (typeof condition === 'boolean' ? condition : condition()),
-    [condition],
-    () => {
-      let initialState: AsyncStateSetterParam<boolean> = {
-        isLoading: true,
-      };
+  const previousCondition = usePrevious(condition);
+
+  const [passed, setPassed] = useState(false);
+  const [isLoading, setLoading] = useState(false);
```

Comment:
Just wondering if this should be named `setIsLoading` to follow the naming convention here?

Reply:
I typically use the `isX` convention just to convey that the variable/state is a boolean. Personally, I think doing this convention with the setter just makes the function sound weird, and it doesn't add a whole lot. We should just be focusing on the name's subject (the X, or 'loading' in this example).

Can change it if you feel particularly strongly about it, but it's not something I'd get others to change personally.

Reply:
Ah, that's good to know. Yeah, I don't have a strong opinion about it (`setLoading` sounds better imo), was mainly wondering if we followed that type of naming convention elsewhere.